posts/lambda-execution-context-demystified.md

```diff
@@ -0,0 +1,128 @@
+This Lambda is exposed via AWS API Gateway and gets executed by the following [endpoint](https://jz2ckwdte0.execute-api.us-east-1.amazonaws.com/Prod/execute/).
+
+```java
+public class LambdaHandler implements RequestHandler<Object, Object> {
+
+    static double STATIC_RANDOM;
+
+    static {
+        STATIC_RANDOM = Math.random(); // (1) Value generated in the static block
+    }
+
+    private final double constructorRandom;
+
+    public LambdaHandler() {
+        this.constructorRandom = Math.random(); // (2) Value generated in the LambdaHandler constructor
+    }
+
+    public Object handleRequest(final Object input, final Context context) {
+        double invocationRandom = Math.random(); // (3) Value generated in the LambdaHandler method for each invocation
+
+        String output = String.format(
+                "{ \"1 - static value\": \"%f\", \" 2 - constructor value\": \"%f\", \"3 - invocation value\": \"%f\"}",
+                STATIC_RANDOM, this.constructorRandom, invocationRandom);
+
+        Map<String, String> headers = new HashMap<>();
+        headers.put("Content-Type", "application/json");
+        headers.put("X-Custom-Header", "application/json");
+        return new GatewayResponse(output, headers, 200);
+    }
+}
+
+```
+
+
+1. The static block is executed when the _Execution Context_ loads the LambdaHandler class in the JVM for the first time. The random value is stored in a static field.
+2. The constructor is executed when the _Execution Context_ instantiates a LambdaHandler instance for the first time. The random value is stored in a field of the instance.
+3. The handleRequest method is executed every time the Lambda function is invoked and the random value stored in a local variable.
+
+
+Accessing the endpoint multiple times returns the following message:
+```json
+{ "1 - static value": "0.936418", " 2 - constructor value": "0.400978", "3 - invocation value": "0.362746"}
+{ "1 - static value": "0.936418", " 2 - constructor value": "0.400978", "3 - invocation value": "0.969463"}
+{ "1 - static value": "0.936418", " 2 - constructor value": "0.400978", "3 - invocation value": "0.0.723602"}
+```
+
+As expected, only the _invocation value_ changes between each invocation. The values generated by both the static block and the constructor remain the same.
+
+If you access the [endpoint](https://jz2ckwdte0.execute-api.us-east-1.amazonaws.com/Prod/execute/) from your browser multiple times, you will notice the same behavior where only the _invocation value_ changes between your requests. The static and constructor values remains the same. The values will probably be different than the example above, that is because AWS has destroyed the _Execution Context_ and a new one had to get created during your first request.
```

Comment:
```suggestion
If you access the [endpoint](https://jz2ckwdte0.execute-api.us-east-1.amazonaws.com/Prod/execute/) from your browser multiple times, you will notice the same behavior where only the _invocation value_ changes between your requests. The static and constructor values remain the same. The values will probably be different than the example above, that is because AWS has destroyed the _Execution Context_ and a new one had to get created during your first request.
```